filecoin-proofs/src/api/update.rs

```diff
@@ -0,0 +1,616 @@
+
+use anyhow::{ensure, Context, Error, Result};
+use bincode::deserialize;
+use blstrs::Scalar as Fr;
+use filecoin_hashers::{Domain, Hasher};
+use fr32::bytes_into_fr;
+use generic_array::typenum::Unsigned;
+use log::info;
+use memmap::MmapOptions;
+use merkletree::merkle::get_merkle_tree_len;
+use merkletree::store::StoreConfig;
+use storage_proofs_core::{
+    cache_key::CacheKey,
+    compound_proof::{self, CompoundProof},
+    merkle::{get_base_tree_count, MerkleTreeTrait},
+    multi_proof::MultiProof,
+    proof::ProofScheme,
+    util::NODE_SIZE,
+};
+use storage_proofs_porep::stacked::{PersistentAux, TemporaryAux, BINARY_ARITY};
+use storage_proofs_update::{
+    constants::TreeRHasher, EmptySectorUpdate, EmptySectorUpdateCompound, PartitionProof,
+    PrivateInputs, PublicInputs, PublicParams, SetupParams,
+};
+
+use crate::{
+    caches::{get_empty_sector_update_params, get_empty_sector_update_verifying_key},
+    constants::{DefaultPieceDomain, DefaultPieceHasher},
+    pieces::verify_pieces,
+    types::{
+        Commitment, EmptySectorUpdateProof, HSelect, PieceInfo, PoRepConfig, UpdateProofPartitions,
+    },
+};
+
+pub fn compare_elements(path1: &Path, path2: &Path) -> Result<(), Error> {
+    info!("Comparing elements between {:?} and {:?}", path1, path2);
+    let f_data1 = OpenOptions::new()
+        .read(true)
+        .open(path1)
+        .with_context(|| format!("could not open path={:?}", path1))?;
+    let data1 = unsafe {
+        MmapOptions::new()
+            .map(&f_data1)
+            .with_context(|| format!("could not mmap path={:?}", path1))
+    }?;
+    let f_data2 = OpenOptions::new()
+        .read(true)
+        .open(path2)
+        .with_context(|| format!("could not open path={:?}", path2))?;
+    let data2 = unsafe {
+        MmapOptions::new()
+            .map(&f_data2)
+            .with_context(|| format!("could not mmap path={:?}", path2))
+    }?;
+    let fr_size = std::mem::size_of::<Fr>() as usize;
+    let end = metadata(path1)?.len() as u64;
+    ensure!(
+        metadata(path2)?.len() as u64 == end,
+        "File sizes must match"
+    );
+
+    for i in (0..end).step_by(fr_size) {
+        let index = i as usize;
+        let fr1 = bytes_into_fr(&data1[index..index + fr_size])?;
+        let fr2 = bytes_into_fr(&data2[index..index + fr_size])?;
+        ensure!(fr1 == fr2, "Data mismatch when comparing elements");
+    }
+    info!("Match found for {:?} and {:?}", path1, path2);
+
+    Ok(())
+}
+
+/// Encodes data into an existing replica.
+/// Returns tuple of (comm_r_new, comm_r_last_new, comm_d_new)
+#[allow(clippy::too_many_arguments)]
+pub fn encode_into<Tree: 'static + MerkleTreeTrait<Hasher = TreeRHasher>>(
+    porep_config: PoRepConfig,
+    new_replica_path: &Path,
+    new_cache_path: &Path,
+    sector_key_path: &Path,
+    sector_key_cache_path: &Path,
+    staged_data_path: &Path,
+    piece_infos: &[PieceInfo],
+) -> Result<(Commitment, Commitment, Commitment)> {
```

Comment:
Please use a struct for this return value, super easy to misuse the API otherwise